src/client/api/extension.d.ts

```diff
@@ -0,0 +1,256 @@
+     * happen either when user doesn't allow this or the extension doesn't allow this.
+     * There are a specific set of extensions that are currently allowed to access this API.
+     */
+    getKernelService(): Promise<IExportedKernelService | undefined>;
+}
+
+/**
+ * The supported Python environment types.
+ */
+export enum EnvironmentType {
+    Unknown = 'Unknown',
+    Conda = 'Conda',
+    VirtualEnv = 'VirtualEnv',
+    Pipenv = 'PipEnv',
+    Pyenv = 'Pyenv',
+    Venv = 'Venv',
+    WindowsStore = 'WindowsStore',
+    Poetry = 'Poetry',
+    VirtualEnvWrapper = 'VirtualEnvWrapper',
+    Global = 'Global',
+    System = 'System'
+}
+
+/**
+ * A representation of a Python runtime's version.
+ */
+export type PythonVersion = {
+    /**
+     * The original version string.
+     */
+    raw: string;
+    major: number;
+    minor: number;
+    patch: number;
+    build: string[];
+    /**
+     * Identifies a tag in the release process (e.g. beta 1)
+     */
+    prerelease: string[];
+};
+export type PythonEnvironment = {
+    displayName?: string;
+    path: string;
+    version?: PythonVersion;
+    sysPrefix: string;
+    envType?: EnvironmentType;
+    envName?: string;
+    envPath?: string;
+};
+
+export interface IJupyterKernelSpec {
+    /**
+     * Id of an existing (active) Kernel from an active session.
+     */
+    id?: string;
+    name: string;
+    language?: string;
+    path: string;
+    env?: NodeJS.ProcessEnv | undefined;
+    /**
+     * Kernel display name.
+     */
+    readonly display_name: string;
+    /**
+     * A dictionary of additional attributes about this kernel; used by clients to aid in kernel selection.
+     * Optionally storing the interpreter information in the metadata (helping extension search for kernels that match an interpreter).
+     */
+    readonly metadata?: Record<string, unknown> & { interpreter?: Partial<PythonEnvironment> };
+    readonly argv: string[];
+    /**
+     * Optionally where this kernel spec json is located on the local FS.
+     */
+    specFile?: string;
+    /**
+     * Optionally the Interpreter this kernel spec belongs to.
+     * You can have kernel specs that are scoped to an interpreter.
+     * E.g. if you have Python in `c:\Python\Python3.8`
+     * Then you could have kernels in `<sys.prefix folder for this interpreter>\share\jupyter\kernels`
+     * Plenty of conda packages ship kernels in this manner (beakerx, java, etc).
+     */
+    interpreterPath?: string;
+    readonly interrupt_mode?: 'message' | 'signal';
+}
+/**
+ * Connection metadata for Kernels started using kernelspec (JSON).
+ * This could be a raw kernel (spec might have path to executable for .NET or the like).
+ * If the executable is not defined in kernelspec json, & it is a Python kernel, then we'll use the provided python interpreter.
+ */
+export type LocalKernelSpecConnectionMetadata = Readonly<{
```

Comment:
I'm confused, this is exported here and in types.ts? Which one do we want to expose to people?

Reply:
Users of this API will need the *.d.ts files (or the types).
This file contains everything in one place, this way they can just get it from here.
Ideally we should push this as an npm package or the like, but we can do that later

Reply:
But doesn't that mean the other location isn't necessary? Can't we just use this definition ourselves?

Reply:
Not really, I don't want to expose all of the properties, I don't want to expose everythign in Python interpreter or some other properties, else we end up in advertently having to support all of the stuff we expose.
This way we have full control over what's exposed.

Reply:
E.g. I don't want to expose everything in the Python interpretre API